torii/src/lib.rs

```diff
@@ -92,171 +95,171 @@ impl Torii {
+                            }
+                        })
+                    )
+                }
+            }))
+            .route(uri::BLOCKS_STREAM,
+                post({
+                    let kura = self.kura.clone();
+                    move |ws: WebSocketUpgrade| {
+                        core::future::ready(
+                            ws.on_upgrade(|ws| async move {
+                                if let Err(error) = routing::handle_blocks_stream(kura, ws).await {
+                                    iroha_logger::error!(%error, "Failed to subscribe to blocks stream");
+                                }
+                            })
+                        )
                     }
                 })
-            });
-
-        let ws_router = events_ws_router.or(blocks_ws_router);
-
-        warp::any()
-            .and(
-                // we want to avoid logging for the "health" endpoint.
-                // we have to place it **first** so that warp's trace will
-                // not log 404 if it doesn't find "/health" which might be placed
-                // **after** `.with(trace)`
-                health_route,
             )
-            .or(ws_router
-                .or(get_router)
-                .or(post_router)
-                .with(warp::trace::request()))
+            ;
+
+        router.layer((
+            TraceLayer::new_for_http()
+                .make_span_with(DefaultMakeSpan::default().include_headers(true)),
+            // Graceful shutdown will wait for outstanding requests to complete.
+            // Add a timeout so requests don't hang forever.
+            TimeoutLayer::new(Duration::from_secs(60)),
+        ))
     }
 
     /// Start main API endpoints.
     ///
     /// # Errors
     /// Can fail due to listening to network or if http server fails
-    fn start_api(self: Arc<Self>) -> eyre::Result<Vec<task::JoinHandle<()>>> {
+    fn start_api(self: Arc<Self>) -> eyre::Result<Vec<task::JoinHandle<eyre::Result<()>>>> {
         let torii_address = &self.address;
 
         let handles = torii_address
             .to_socket_addrs()?
             .map(|addr| {
                 let torii = Arc::clone(&self);
-
                 let api_router = torii.create_api_router();
-                let signal_fut = async move { torii.notify_shutdown.notified().await };
-                // FIXME: warp panics if fails to bind!
-                //        handle this properly, report address origin after Axum
-                //        migration: https://github.com/hyperledger/iroha/issues/3776
-                let (_, serve_fut) =
-                    warp::serve(api_router).bind_with_graceful_shutdown(addr, signal_fut);
 
-                task::spawn(serve_fut)
+                let signal = async move { torii.notify_shutdown.notified().await };
+
+                let serve_fut = async move {
+                    let listener = TcpListener::bind(addr).await?;
```

Comment:
Here, it is possible to attach the configuration source of the address to the error.

Reply:
What do you mean exactly by "attach the configuration source of the address"? And how to do that?

Reply:
I mean to return a report with an information on where the `addr` came from. For example:

https://github.com/hyperledger/iroha/blob/056cae387d51d120264facaf8bc0c13c1d671784/cli/src/lib.rs#L662-L668

Reply:
Fixed, `torii` now exit early if fails to bind address

```plain
Error: Could not start Iroha
│
├─▶ Unable to start Torii (Iroha HTTP API Gateway)
│
╰─▶ Address already in use (os error 48)
    ╰╴config origin: parameter `torii.address` with value `127.0.0.1:8080` in file `test/peers/iroha0/config.toml`
```